src/main/scala/com/wavesplatform/history/HistoryWriterImpl.scala

```diff
@@ -55,42 +53,41 @@ class HistoryWriterImpl private(file: Option[File], val synchronizationToken: Re
   override def discardBlock(): Seq[Transaction] = write { implicit lock =>
     val h = height()
     val transactions =
-      Block.parseBytes(blockBodyByHeight.mutate(_.remove(h))).fold(_ => Seq.empty[Transaction], _.transactionData.map(_.transaction))
-    scoreByHeight.mutate(_.remove(h))
-    val vOpt = Option(blockIdByHeight.mutate(_.remove(h)))
-    vOpt.map(v => heightByBlockId.mutate(_.remove(v)))
-    db.commit()
+      Block.parseBytes(blockBytes(h).getOrElse(Array[Byte]())).fold(_ => Seq.empty[Transaction], _.transactionData.map(_.transaction))
+
+    delete(blockBodyByHeightKey(h), None)
+    delete(scoreByHeightKey(h), None)
+    get(blockBodyByHeightKey(h)).foreach(b => ByteStrCodec.decode(b).toOption.foreach(r => delete(heightByBlockIdKey(r.value), None)))
+    delete(blockIdByHeightKey(h), None)
+    put(heightKey, Ints.toByteArray(h - 1), None)
 
     transactions
   }
 
 
   override def lastBlockIds(howMany: Int): Seq[ByteStr] = read { implicit lock =>
-    (Math.max(1, height() - howMany + 1) to height()).flatMap(i => Option(blockIdByHeight().get(i)))
+    (Math.max(1, height() - howMany + 1) to height()).flatMap(i => 
+      get(blockIdByHeightKey(i))
+      .flatMap(b => ByteStrCodec.decode(b).toOption.map(r => r.value)))
       .reverse
   }
 
-  override def height(): Int = read { implicit lock => blockIdByHeight().size() }
+  override def height(): Int = read { implicit lock => 
+    get(heightKey).map(Ints.fromByteArray).get
+  }
```

Comment:
Does it better to record `height` in db than check the size in original way?

Reply:
leveldb does not support the ```count``` operation. So an independent counter is required.

Reply:
i see